File: ml/merchant_normalize.py
"""
Merchant normalization: reduce noise in transaction descriptions.
Python port of lib/nlp/merchant-normalize.ts for use by the classifier.
"""

from __future__ import annotations

import json
import re
from pathlib import Path

def _fuzzy_best_match(text: str, candidates: list[str]) -> tuple[str, float] | None:
    """Return (best_match, score) or None. Uses rapidfuzz if available, else simple Jaccard."""
    try:
        from rapidfuzz import fuzz
        best = max(candidates, key=lambda c: fuzz.ratio(text, c))
        score = fuzz.ratio(text, best) / 100.0
        return (best, score) if score >= 0.85 else None
    except ImportError:
        # Fallback: Jaccard similarity on character bigrams
        def jaccard(a: str, b: str) -> float:
            sa = set(a[i:i+2] for i in range(len(a)-1))
            sb = set(b[i:i+2] for i in range(len(b)-1))
            if not sa and not sb:
                return 1.0
            inter = len(sa & sb)
            union = len(sa | sb)
            return inter / union if union else 0.0
        if not candidates:
            return None
        best = max(candidates, key=lambda c: jaccard(text, c))
        score = jaccard(text, best)
        return (best, score) if score >= 0.85 else None

# Built-in variant map (matches merchant-normalize.ts + synthetic data brands)
BUILTIN_VARIANTS: dict[str, list[str]] = {
    "mcdonalds": ["mcd", "mcds", "mcdonalds", "macdonalds"],
    "starbucks": ["starbucks", "starbucks coffee", "sbux"],
    "walmart": ["walmart", "wal mart", "wm supercenter"],
    "target": ["target", "target store"],
    "costco": ["costco", "costco wholesale"],
    "amazon": ["amazon", "amazon.com", "amzn"],
    "shell": ["shell", "shell oil"],
    "exxon": ["exxon", "exxonmobil", "exxon mobil"],
    "chevron": ["chevron", "chevron texaco"],
    "subway": ["subway", "subway sandwiches"],
    "chipotle": ["chipotle", "chipotle mexican grill"],
    "whole foods": ["whole foods", "whole foods market"],
    "trader joes": ["trader joes", "trader joe", "trader joes market"],
    "uber": ["uber", "uber trip", "uber eats"],
    "lyft": ["lyft", "lyft ride"],
    "venmo": ["venmo", "venmo payment"],
    "spotify": ["spotify", "spotify usa"],
    "netflix": ["netflix", "netflix.com"],
    "hulu": ["hulu", "hulu llc"],
    "apple": ["apple", "apple.com", "apple store", "itunes"],
    "pepsi": ["pepsi", "pepsi bottling", "pepsi co", "ctlppepsi", "ctlpepsi", "ctlp pepsi"],
    "mta": ["mta", "mta nyct", "mta paygo", "mta mnr", "mta lirr"],
    "njt": ["njt", "njt rail", "njt rail my tix", "nj transit"],
    "dominos": ["dominos", "dominos pizza", "domino"],
    "pizza hut": ["pizza hut", "pizza hut delivery"],
    "dunkin": ["dunkin", "dunkin donuts", "dunkin coffee"],
    "panera": ["panera", "panera bread"],
    "home depot": ["home depot", "the home depot"],
    "lowes": ["lowes", "lowes home improvement"],
    "best buy": ["best buy", "bestbuy"],
    "cvs": ["cvs", "cvs pharmacy", "cvs caremark"],
    "walgreens": ["walgreens", "walgreens pharmacy"],
    "safeway": ["safeway", "safeway inc"],
    "kroger": ["kroger", "kings supers", "ralphs", "fred meyer"],
    "sprouts": ["sprouts", "sprouts farmers market"],
}

STATE_ABBREVS = frozenset(
    "al ak az ar ca co ct de fl ga hi id il in ia ks ky la me md ma mi mn ms "
    "mo mt ne nv nh nj nm ny nc nd oh ok or pa ri sc sd tn tx ut vt va wa wv wi wy dc".split()
)

CITY_NAMES = frozenset(
    "new york los angeles san francisco san diego las vegas newark raleigh "
    "chicago houston phoenix philadelphia boston seattle denver austin miami atlanta".split()
)

TLDS = frozenset(["com", "net", "org", "co", "io"])

_variant_map: dict[str, str] = {}
_single_word_map: dict[str, str] = {}


def _rebuild_maps(variants: dict[str, list[str]] | None = None) -> None:
    global _variant_map, _single_word_map
    v = variants or BUILTIN_VARIANTS
    _variant_map = {}
    _single_word_map = {}
    for canonical, aliases in v.items():
        for alias in aliases:
            key = alias.strip().lower().replace(" ", "_")
            _variant_map[key] = canonical.replace("_", " ")
            token = alias.split()[0].lower() if alias.split() else ""
            if token and token not in _single_word_map:
                _single_word_map[token] = canonical.replace("_", " ")
    for canonical in v:
        key = canonical.replace("_", " ").replace(" ", "_")
        _variant_map[key] = canonical.replace("_", " ")


def load_merchant_variants(path: Path | str) -> None:
    """Load and merge variants from JSON file."""
    p = Path(path)
    if not p.exists():
        return
    data = json.loads(p.read_text())
    merged = dict(BUILTIN_VARIANTS)
    for k, v in data.items():
        if k.startswith("_"):
            continue
        merged[k] = v
    _rebuild_maps(merged)


_rebuild_maps()
# Load data/merchant-variants.json if present
_proj_root = Path(__file__).resolve().parent.parent
load_merchant_variants(_proj_root / "data" / "merchant-variants.json")


def _remove_statement_boilerplate(text: str) -> str:
    t = re.sub(r"^purchase\s+authorized\s+on\s+\d{1,2}\s*/?\s*\d{1,2}\s*", "", text, flags=re.I)
    t = re.sub(r"^payment\s+", "", t, flags=re.I)
    return t.strip()


def _remove_store_numbers(text: str) -> str:
    t = re.sub(r"#\s*\d+", "", text)
    t = re.sub(r"\bstore\s*\d+\b", "", t, flags=re.I)
    t = re.sub(r"\bunit\s*\d+\b", "", t, flags=re.I)
    return t


def _remove_locations(text: str) -> str:
    words = text.split()
    filtered = []
    for w in words:
        lw = w.lower()
        if len(lw) == 2 and lw in STATE_ABBREVS:
            continue
        filtered.append(w)
    return " ".join(filtered)


def _remove_city_names(text: str) -> str:
    if not text.strip():
        return ""
    result = text.strip().lower()
    words = result.split()
    no_states = [w for w in words if not (len(w) == 2 and w in STATE_ABBREVS)]
    result = " ".join(no_states)
    for city in CITY_NAMES:
        pattern = r"\s+" + re.escape(city) + r"\s*$"
        result = re.sub(pattern, "", result, flags=re.I)
        result = re.sub(r"\s+", " ", result).strip()
    return result


def _strip_tlds(text: str) -> str:
    words = text.split()
    return " ".join(w for w in words if w.lower() not in TLDS)


def _apply_merchant_variants(text: str) -> str | None:
    trimmed = re.sub(r"\s+", " ", text.strip().lower())
    words = [w for w in trimmed.split() if w]

    for length in range(min(len(words), 4), 0, -1):
        phrase = "_".join(words[:length])
        if phrase in _variant_map:
            return _variant_map[phrase]

    for word in words:
        if word in _single_word_map:
            return _single_word_map[word]

    if _variant_map:
        phrase = trimmed.replace(" ", "_")
        variant_keys = list(_variant_map.keys())
        match = _fuzzy_best_match(phrase, variant_keys)
        if match and match[1] >= 0.85:
            return _variant_map[match[0]]

    return None


def _first_n_words(text: str, n: int = 3) -> str:
    words = text.split()
    return " ".join(words[:n])


def normalize_merchant(raw_description: str) -> str:
    """
    Normalize raw transaction description to canonical merchant name.
    Strips boilerplate, IDs, locations; applies variant mapping; uses first-N-words for unknown.
    """
    if not raw_description or not str(raw_description).strip():
        return ""

    t = str(raw_description).lower().strip()
    t = re.sub(r"[^\w\s#]", " ", t)
    t = re.sub(r"\b[s]?\d{10,}\b", "", t, flags=re.I)
    t = re.sub(r"\bcard\s*\d{4}\b", "", t, flags=re.I)
    t = re.sub(r"\b\d{3}[-.]?\d{3}[-.]?\d{4}\b", "", t)
    t = re.sub(r"\s+", " ", t).strip()

    t = _remove_statement_boilerplate(t)
    t = _remove_store_numbers(t)
    t = _remove_locations(t)
    t = _remove_city_names(t)
    t = _strip_tlds(t)
    t = re.sub(r"\s+", " ", t).strip()

    canonical = _apply_merchant_variants(t)
    if canonical is not None:
        return canonical

    return _first_n_words(t, 3)

File: lib/nlp/merchant-normalize.ts
/**
 * Merchant normalization: reduce noise in transaction descriptions.
 * - Remove store numbers (#1234), locations (NY, NYC, city names)
 * - Standardize common merchant name variants (from built-in or external map)
 * - Fuzzy matching for variant lookup and repeated merchants
 * - First-N-words heuristic for unknown merchants
 */

import stringSimilarity from "string-similarity";
import { removeCityNames, firstNWords, stripTlds } from "./enhancements";

export type MerchantVariantMap = Record<string, string[]>;

/** Built-in seed. Extend via data/merchant-variants.json or setMerchantVariants(). */
const BUILTIN_VARIANTS: MerchantVariantMap = {
  mcdonalds: ["mcd", "mcds", "mcdonalds", "macdonalds"],
  starbucks: ["starbucks", "starbucks coffee", "sbux"],
  walmart: ["walmart", "wal mart", "wm supercenter"],
  target: ["target", "target store"],
  costco: ["costco", "costco wholesale"],
  amazon: ["amazon", "amazon.com", "amzn"],
  shell: ["shell", "shell oil"],
  exxon: ["exxon", "exxonmobil", "exxon mobil"],
  chevron: ["chevron", "chevron texaco"],
  subway: ["subway", "subway sandwiches"],
  chipotle: ["chipotle", "chipotle mexican grill"],
  whole_foods: ["whole foods", "whole foods market"],
  trader_joes: ["trader joes", "trader joe", "trader joes market"],
  uber: ["uber", "uber trip", "uber eats"],
  lyft: ["lyft", "lyft ride"],
  venmo: ["venmo", "venmo payment"],
  spotify: ["spotify", "spotify usa"],
  netflix: ["netflix", "netflix.com"],
  hulu: ["hulu", "hulu llc"],
  apple: ["apple", "apple.com", "apple store", "itunes"],
  pepsi: ["pepsi", "pepsi bottling", "pepsi co", "ctlppepsi", "ctlpepsi"],
  mta: ["mta", "mta nyct", "mta paygo", "mta mnr", "mta lirr"],
  njt: ["njt", "njt rail", "nj transit"],
};

// Mutable variant map: merge built-in + any loaded custom data
let activeVariantMap = new Map<string, string>();
let activeSingleWordMap = new Map<string, string>();

function rebuildMaps(variants: MerchantVariantMap) {
  const variantMap = new Map<string, string>();
  const singleWordMap = new Map<string, string>();
  for (const [canonical, list] of Object.entries(variants)) {
    for (const v of list) {
      const key = v.replace(/\s+/g, " ").trim().replace(/\s+/g, "_");
      variantMap.set(key, canonical);
      const token = v.split(/\s+/)[0];
      if (token && !singleWordMap.has(token)) singleWordMap.set(token, canonical);
    }
    variantMap.set(canonical.replace(/_/g, " "), canonical);
  }
  activeVariantMap = variantMap;
  activeSingleWordMap = singleWordMap;
}

rebuildMaps(BUILTIN_VARIANTS);

/**
 * Replace or extend the merchant variant map.
 * Use to load a larger external map (JSON file, API, etc.).
 * Merges with built-in unless replaceBuiltin is true.
 */
export function setMerchantVariants(
  variants: MerchantVariantMap,
  replaceBuiltin = false
): void {
  const merged = replaceBuiltin ? variants : { ...BUILTIN_VARIANTS, ...variants };
  rebuildMaps(merged);
}

/**
 * Load variant map from JSON. Format: { "canonical": ["variant1", "variant2"], ... }
 */
export async function loadMerchantVariantsFromUrl(url: string): Promise<void> {
  const res = await fetch(url);
  const data = (await res.json()) as MerchantVariantMap;
  setMerchantVariants(data, false);
}

// US state abbreviations (2-letter)
const STATE_ABBREVS = new Set([
  "al", "ak", "az", "ar", "ca", "co", "ct", "de", "fl", "ga",
  "hi", "id", "il", "in", "ia", "ks", "ky", "la", "me", "md",
  "ma", "mi", "mn", "ms", "mo", "mt", "ne", "nv", "nh", "nj",
  "nm", "ny", "nc", "nd", "oh", "ok", "or", "pa", "ri", "sc",
  "sd", "tn", "tx", "ut", "vt", "va", "wa", "wv", "wi", "wy", "dc"
]);

const LOCATION_ABBREVS = new Set([
  "ny", "nyc", "la", "sf", "chicago", "houston", "phoenix", "philly",
  "dc", "nj", "tx", "nc", "fl"
]);

/** Strip Wells Fargo boilerplate: "purchase authorized on 11 23" or "purchase authorized on 11/23" */
function removeStatementBoilerplate(text: string): string {
  return text
    .replace(/^purchase\s+authorized\s+on\s+\d{1,2}\s*\/?\s*\d{1,2}\s*/i, "")
    .replace(/^payment\s+/i, "")
    .trim();
}

function removeStoreNumbers(text: string): string {
  return text
    .replace(/#\s*\d+/g, "")
    .replace(/\bstore\s*\d+\b/gi, "")
    .replace(/\bunit\s*\d+\b/gi, "")
    .replace(/\b\d+\s*(?:st|nd|rd|th)\s+(?:ave|street|st)\b/gi, "");
}

function removeLocations(text: string): string {
  const words = text.split(/\s+/);
  const filtered: string[] = [];
  for (let i = 0; i < words.length; i++) {
    const w = words[i].toLowerCase();
    if (w.length === 2 && STATE_ABBREVS.has(w)) continue;
    if (LOCATION_ABBREVS.has(w)) continue;
    filtered.push(words[i]);
  }
  return filtered.join(" ");
}

function applyMerchantVariants(text: string): string | null {
  const trimmed = text.replace(/\s+/g, " ").trim().toLowerCase();
  const words = trimmed.split(/\s+/).filter(Boolean);

  for (let len = Math.min(words.length, 4); len >= 1; len--) {
    const phrase = words.slice(0, len).join("_");
    const canonical = activeVariantMap.get(phrase);
    if (canonical) return canonical.replace(/_/g, " ");
  }

  for (const word of words) {
    const canonical = activeSingleWordMap.get(word);
    if (canonical) return canonical.replace(/_/g, " ");
  }

  // Fuzzy match against known variants
  const variantStrings = Array.from(activeVariantMap.keys());
  const match = stringSimilarity.findBestMatch(trimmed.replace(/\s+/g, "_"), variantStrings);
  if (match.bestMatch.rating >= 0.85) {
    const canonical = activeVariantMap.get(match.bestMatch.target);
    if (canonical) return canonical.replace(/_/g, " ");
  }

  return null; // no match → caller uses firstNWords heuristic
}

/**
 * Full merchant normalization pipeline.
 */
export function normalizeMerchant(text: string): string {
  if (!text?.trim()) return "";

  let result = text
    .toLowerCase()
    .replace(/[^\w\s#]/g, " ")
    .replace(/\s+/g, " ")
    .trim();

  result = removeStatementBoilerplate(result);
  result = removeStoreNumbers(result);
  result = removeLocations(result);
  result = removeCityNames(result);
  result = stripTlds(result);
  result = result.replace(/\s+/g, " ").trim();

  const canonical = applyMerchantVariants(result);
  if (canonical != null) return canonical;

  // First-N-words heuristic for unknown merchants (max 3 to preserve org names)
  return firstNWords(result, 3);
}

/**
 * Normalize merchant for categorization.
 */
export function normalizeMerchantDescription(rawDescription: string): string {
  const cleaned = rawDescription
    .toLowerCase()
    .replace(/[^\w\s#]/g, " ")
    .replace(/\b[s]?\d{10,}\b/gi, "")
    .replace(/\bcard\s*\d{4}\b/gi, "")
    .replace(/\b\d{3}[-.]?\d{3}[-.]?\d{4}\b/g, "")
    .replace(/\s+/g, " ")
    .trim();

  return normalizeMerchant(cleaned);
}

/**
 * Fuzzy similarity (Jaccard on word tokens). Returns 0–1.
 */
export function merchantSimilarity(a: string, b: string): number {
  const wordsA = new Set(a.toLowerCase().split(/\s+/).filter(Boolean));
  const wordsB = new Set(b.toLowerCase().split(/\s+/).filter(Boolean));
  if (wordsA.size === 0 && wordsB.size === 0) return 1;
  if (wordsA.size === 0 || wordsB.size === 0) return 0;
  const intersection = [...wordsA].filter((w) => wordsB.has(w)).length;
  const union = new Set([...wordsA, ...wordsB]).size;
  return intersection / union;
}
